cranelift-codegen/meta/src/isa/x86/instructions.rs

```diff
@@ -249,5 +249,32 @@ pub fn define(
         .operands_out(vec![y, rflags]),
     );
 
+    let TxN = &TypeVar::new(
+        "TxN",
+        "A SIMD vector type",
+        TypeSetBuilder::new()
+            .ints(Interval::All)
+            .floats(Interval::All)
+            .bools(Interval::All)
+            .simd_lanes(Interval::All)
+            .includes_scalars(false)
+            .build(),
+    );
+    let a = &operand("a", TxN);
+    //let b = &operand("b", TxN);
+    ig.push(
+        Inst::new(
+            "x86_pshuf",
+            r#"
+    Shuffle Packed -- copies data from either memory or lanes in an extended
+    register and re-orders the data according to the passed immediate byte.
+    TODO allow memory type here
+    TODO how to allow passing in immediate during code generation?
+    "#,
+        )
+            .operands_in(vec![a])
+            .operands_out(vec![a]),
+    );
```

Comment:
I thought it would be most transparent to future engineers if I just implement the x86 instruction itself so they know what is intended; however, for this to be truly flexible I would like to be able to pass in the `u8` immediate like `x86_pshuf.bind_vector(I32, 4).set_immediate(0x...)` in the encodings. Obviously `set_immediate` does not exist but I see a `rrr` used to pass in constants to the template; is that only for register addressing?

Reply:
Never mind, I think I see how to do this: add a new operand here based on teh `immediates.by_name("uimm8")` and then in `legalize.rs` do something like:
```
    let shuffle_encoding = Literal::constant(uimm8, 0x00);
    group.legalize(
        def!(y = splat(x)),
        vec![
            def!(a = bitcast(x)),
            def!(y = x86_pshuf(a, shuffle_encoding))
        ]
    );
```
Does that sound right?

Reply:
Yes indeed. This should also allow to transform pshuf with specific immediates into smaller / faster sequences of instructions, in a (future) pipehole optimizer.